Sources/protoc-gen-swift/MessageGenerator.swift

```diff
@@ -506,6 +506,23 @@ fileprivate func hasSingleMessageField(descriptor: Descriptor) -> Bool {
   return result
 }
 
+fileprivate func hasRecursiveSingularField(descriptor: Descriptor, visited: [Descriptor] = []) -> Bool {
+  var visited = visited
```

Comment:
If you don't default `visited`, and make it `inout`, can you ensure every node is only checked once?  i.e. - since it is a copy, the deeper nodes can get visited multiple times because nothing is shared between those sub visits?

Reply:
@thomasvl I guess it could be a `Set<Descriptor>`. As it stands, it short-circuits on any recursion, but doesn’t share state between branches. Are you thinking about this as an optimization?

Reply:
Is there a straightforward way to benchmark `protoc` for the test protos and compare against `master`? I guess that would answer the question.

Reply:
@thomasvl tried `inout visited` and got 19 additional heap-allocated messages:

https://github.com/ydnar/swift-protobuf/commit/11698e3cf60f434646e0e3a808e689fcac26fd8a

Reply:
I'll have to digest this one a little more, I didn't expect there to be any change in what it decided should use the heap.

Reply:
I think maybe I see the logic issue; `visited` is sorta doing two things.  It is tracking what we've already checked, but also trying to check for recursion, and the second part isn't completely correct.

```protobuf
message A {
  string name = 1;
  B b = 2;
}

message B {
  C c = 1;
}

message C {
  D d = 1;
}

message D {
  B b = 1;
}
```

Shouldn't `A` not need storage because it isn't recursive itself.  But `B`, `C`, and `D` all should use storage because they define a recursive sub loop?

Since `visited` is serving double duty for tracking what's been checked and if there is recursion, it can end up saying that `A` has to use storage?  So the algorithm needs to check purely if the root we are starting from is used again by a sub field, but something like `visitited` is needed to not get stuck in a sub cycle that doesn't directly impact the decision for the root message.

Reply:
To your other questions - yes I'm concerned about performance here when the message graph gets deep/wide and not doing to much extra work.  As far as benchmarking it, the easiest way is likely to just time making the reference files, but all the files are so impl time times will likely be dwarfed by the other parts of the the overhead.

Reply:
@thomasvl thanks for discovering the logic issue. I think you’re right in your example, `A` won’t need heap storage, but `B` will.

I had originally coded this without a `visited` array, just checking if the root message (in this case `A`) was referenced by any sub-messages. I’ll try that again.

Unrelated: on my machine (2018 MB Air), using Swift 5.1, the package fails to compile because of a missing `default` statement in a generated `switch` (`generated_swift_names_enum_cases.pb.swift:2605`). I have to manually insert it (and accept the inevitable warnings) in order to get the package to build and tests to run.

Reply:
@thomasvl I revised the logic based on your findings, and the result is several fewer heap allocations: db85e488

Reply:
Since it is hard to tell from the diffs, what is left using storage, and does it make sense?

Reply:
The remaining heap allocations look like they makes sense. I’d love to see it stress-tested by someone hitting performance limits with heap allocs like #733.

Reply:
Solving this optimally is tricky:  In @thomasvl's example above, we technically only _need_ storage for one of `B`, `C`, and `D`.  You don't need all three.

Some quick digging suggests that the optimal answer requires solving something called the "maximal acyclic subgraph" problem.  That sounds like more work than I really want to tackle in this PR.